### PR TITLE
Add MariaDB addon

### PR DESCRIPTION
**Pull request by BanzaiMan**
This PR adds MariaDB addon, based on instructions given at https://downloads.mariadb.org/mariadb/repositories/#mirror=digitalocean-nyc&distro=Ubuntu&distro_release=precise&version=10.0.

Versions 5.5, 10.0, and 10.1 can be specified in `.travis.yml`; e.g.,

```yaml
addons:
  mariadb: "10.0"
```

Note that MariaDB is installed with the package-provided configuration file `-o Dpkg::Options::='--force-confnew'`, so that it is run on disk, not ramfs, so it can be potentially slow.

User `travis` needs to be set up, but for some reason `root` cannot.

```sh-session
travis@debug-BanzaiMan-playing-around-php-1412545209:~$ mysql -u root 
Welcome to the MariaDB monitor.  Commands end with ; or \g.
Your MariaDB connection id is 252
Server version: 10.0.14-MariaDB-1~precise-log mariadb.org binary distribution

Copyright (c) 2000, 2014, Oracle, SkySQL Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [(none)]> CREATE USER 'travis'@'%' IDENTIFIED BY '';          GRANT ALL PRIVILEGES ON *.* TO 'travis'@'%';
ERROR 1396 (HY000): Operation CREATE USER failed for 'travis'@'%'
Query OK, 0 rows affected (0.00 sec)
```

### CODE REVIEW

**Comment by grooverdan** [comment]
' CREATE USER 'travis'@'%' IDENTIFIED BY '';' failing would indicate that the travis user already exists supported by the fact the grant succeeded.

What does 'but for some reason root cannot.' mean? it appears you have a root user. Can you show select user,host,password from mysql.user and show grants in the test buiild?

**Comment by grooverdan** [comment]
also mysql_upgrade might be running in the background (/etc/mysql/debian-start) meaning some of the system table structure may not be fully ready for changes.

**Comment by BanzaiMan** [comment]
Picked this up again, and it *seems* to work with my limited testing.

https://staging.travis-ci.org/BanzaiMan/travis_staging_test/builds/433699#L205 (10.0.20)
https://staging.travis-ci.org/BanzaiMan/travis_staging_test/builds/433701#L214 (10.1.5)

I don't have an app to run it through more rigorous testing, though.

I do not like it that the Ubuntu release code name is hard coded at the moment.

**Comment by grooverdan** [comment]
Provided the following return the right user and a mariadb version it should be ok
<pre>
mysql -u root -e 'SELECT VERSION(); SELECT CURRENT_USER();'
mysql -u travis -e 'SELECT VERSION(); SELECT CURRENT_USER();'
</pre>

> I do not like it that the Ubuntu release code name is hard coded at the moment.

<pre>sh.cmd "add-apt-repository 'deb http://#{MARIADB_MIRROR}/mariadb/repo/#{mariadb_version}/ubuntu ${TRAVIS_UBUNTU_RELEASE} main'", sudo: true
</pre>
TRAVIS_UBUNTU_RELEASE is in a single quote when run so its not interpolated.

**Comment by BanzaiMan** [comment]
@grooverdan Yay. User creds seem to work. https://staging.travis-ci.org/BanzaiMan/travis_staging_test/builds/433703

I'll revisit the release code name tomorrow.

**Comment by grooverdan** [comment]
thanks for revisiting this. Looks good.

**Comment by BanzaiMan** [comment]
Cool. it worked.

**Comment by BanzaiMan** [comment]
https://staging.travis-ci.org/BanzaiMan/travis_staging_test/builds/433803#L221

**Comment by BanzaiMan** [comment]
By rough accounting, it adds about 30 seconds to the build.

**Comment by joshk** [comment]
@henrikhodne @meatballhat can you review the PR pretty please

File: lib/travis/build/addons/mariadb.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+require 'travis/build/addons/base'
+
+module Travis
+  module Build
+    class Addons
+      class Mariadb < Base
+        SUPER_USER_SAFE = true
+
+        MARIADB_GPG_KEY = '0xcbcb082a1bb943db'
+        MARIADB_MIRROR  = 'nyc2.mirrors.digitalocean.com'
```
Comment (BanzaiMan) [inline lib/travis/build/addons/mariadb.rb]: Not clear if this is the best mirror.

https://downloads.mariadb.org/mariadb/repositories/ shows James Madison University, which is physically closer to VA data centers.

**Comment by BanzaiMan** [comment]
@grooverdan Hello! We have deployed this to production (both .org and .com; not yet documented). Could you test this and see if it works to your satisfaction? We would like to get documentation ready, very much like what we have for [MySQL](http://docs.travis-ci.com/user/database-setup/#MySQL). Could you review the documentation, and see if there are any notable differences between MySQL and MariaDB? (Please open a separate issue if you find any!)

Thank you for your help!

**Comment by grooverdan** [comment]
@BanzaiMan :+1: thank you

basic support on [5.5](https://travis-ci.org/grooverdan/PyMySQL/builds/72040993), [10.0](https://travis-ci.org/grooverdan/PyMySQL/builds/72041230) and [10.1](https://travis-ci.org/grooverdan/PyMySQL/jobs/72041107) looks great.

[single list form](https://github.com/grooverdan/PyMySQL/commit/f70e40fe3b5e1fe406f19f5f1fb3777fb62712a6) resulted in [[5.5] as the version](https://travis-ci.org/grooverdan/PyMySQL/jobs/72040844) and hence didn't install 5.5.

[multiple list form](https://github.com/grooverdan/PyMySQL/commit/038d47a5a58d069eb22f895b89142a02e2a34f29) seemed to [ignore all repos](https://travis-ci.org/grooverdan/PyMySQL/jobs/72041918) and resort to mysql.

My [guess on a matrix form](https://github.com/grooverdan/PyMySQL/commit/cda4193dbefc127e9db814bb7b3bcfc7b25131c9) was [silently ignored](https://travis-ci.org/grooverdan/PyMySQL/builds/72040687).

A multiple list form would be really nice to test an application against multiple versions.

Is having *none* as a item in a list acceptable to revert to the installed MySQL so an application can be test against mariadb versions and mysql.

Minor note, Is the exclude/include syntax of travis-ci/travis-ci/issues/1492 supported somehow? It will be useful when multiple MySQL versions to exclude a matrix where MariaDB and MySQL versions get installed.

For container builds having /etc/mysql writeable would be useful - its currently root owned. To protect against users running items as root a check will need to be done to ensure there isn't *user=* set in a *[mysqld_safe]* or *[safe_mysqld]* section (its the OS user that mysqld will start at and by pushing user plugins they could effectively escalate to root). Look for --user= in the output of <pre>my_print_defaults --loose-verbose mysqld_safe safe_mysqld mariadb_safe</pre>. Using the MySQL version of my_print_defaults is sufficient for this test.

Is there a stage where custom config files can be pushed into /etc/mysql/ so they are there when the package is installed and hence the service started? Its quite possible to install config files there that will affect mariadb and not mysql (putting config files having sections *[mariadb]* *[mariadb-10.0]*, *[mariadb-5.5]*, *[mariadb-10.1]* (which is something that would be useful to document).

Exporting in an environment variable which mariadb version was installed would enable a user to do specific scripts/actions/configs/tests for mariadb that might not be accessible on MySQL (or vise versa).

Apart from the minor tip on introducing mariadb specific configuration I can't see anything that would be useful to include in travis documentation.

Sorry if this contains a little too many feature requests. Its a great capability to start with so thank you.

**Comment by grooverdan** [comment]
think I worked out to do [matrix builds](https://github.com/grooverdan/PyMySQL/blob/5c0ef86a38aaf938973e4cc00c3b03bfdfa52ab0/.travis.yml), [not perfect](https://travis-ci.org/grooverdan/PyMySQL/builds/72078087) but getting closer

File: lib/travis/build/addons/mariadb.rb (hunk 1 above)
Comment (michael-berlin) [inline lib/travis/build/addons/mariadb.rb]: Out of curiosity I did a quick and dirty benchmark: https://travis-ci.org/youtube/vitess/builds/72222979#L668

nyc2 from Digital Ocean is the fastest mirror :-) Up to 50 MB/s compared to 3.5MB/s from James Madison University :)

**Comment by grooverdan** [comment]
@BanzaiMan  did you want the above broken up to separate feature requests?

**Comment by BanzaiMan** [comment]
@grooverdan Yes, please. This is a PR, and feature requests tend to be lost here.

**Comment by grooverdan** [comment]
#4552  #4553  #4554  #4555 refers. Thanks again.

**Comment by ntwb** [comment]
Just wanted to say thanks, I've started testing this with WordPress, works as expected :+1:

**Pull request merged**